Implement Callback API

jacobcpeters opened the pull request:
Creating a PR for review. This code should take care of Issue #2. 

@musocrat if you could take a look, I would be happy to work on any problems you can find. I would like an extra pair of eyes, and someone else to test this before I move forward.

jacobcpeters committed:
Merge pull request #2 from jacobcpeters/input-functions

Input functions

jacobcpeters committed:
Merge pull request #3 from JsAudioOrg/master

Merge with main fork

jacobcpeters committed:
Merge pull request #8 from jacobcpeters/master

Merge with main fork

jacobcpeters committed:
Revert "Merge with main fork"

jacobcpeters committed:
Merge pull request #9 from JsAudioOrg/revert-8-master

Revert incorrect pull request

jacobcpeters committed:
Merge pull request #4 from JsAudioOrg/master

Merge with main fork

jacobcpeters committed:
Initial implementation of JsPaStreamCallback class

Note that this is not a direct implementation of the PaStreamCallback type. Instead it is a class to facilitate the cross thread communication between the portaudio thread and the node event thread.

jacobcpeters committed:
Added callback code to PaOpenDefaultStream

jacobcpeters committed:
Created a test for the callback

jacobcpeters committed:
Removed debugging code

jacobcpeters committed:
Added callback code to PaOpenStream

jacobcpeters committed:
Renamed JsPaStreamCallback to JsPaStreamCallbackBridge

This will prevent anyone from misunderstanding what the class is really doing.

jacobcpeters committed:
Further Implement JsPaStreamCallbackBridge

jacobcpeters committed:
Update intermediary PortAudio callback

jacobcpeters committed:
Add sine output to test script

jacobcpeters committed:
Refactoring of callback test into example

jacobcpeters committed:
Cleaned up example code

jacobcpeters committed:
Refactored CallbackBridge to work with more sample formats

jacobcpeters committed:
Added comments to ReadStream and WriteStream to clarify code

The only thing that Nan::TypedArrayContents uses the template types for are the length attribute and the pointer type. Both are irrelevant in the current use case. The length attribute isn't used, and the PAReadStream and PaWriteStream functions both use void pointers as parameters.

jacobcpeters committed:
Removed unused variable

jacobcpeters committed:
Added a callback wire example

jacobcpeters committed:
Refactor JsPaStreamCallbackBridge to accept a userData Object

jacobcpeters committed:
Call deconstructors when stream is stopped, and add support for userData Object

jacobcpeters committed:
Refactor example to show how to use a userData Object

jacobcpeters committed:
Added a few comments

musocrat commented:
For sure, I'm excited to test it out. Just glancing over the code it looks like an elegant solution to the callback problem. I'll dig in tomorrow and do some testing then get back with you. Thanks

musocrat commented:
It's working great for me and the code looks good. I think it's good to merge. That being said, this pull request is against the callback branch which was divergent from master and has nothing of any use. Perhaps make the pull request against master as there should be no conflicts and we can get rid of the callback branch here.

jacobcpeters committed:
Make Js callback synchronous with PortAudio callback

This greatly reduces memory operations, and negates the need for a mutex. It will eliminate the need for extra, complicated, code to implement paFramesPerBufferUnspecified.

jacobcpeters commented:
I just pushed a commit that greatly reduces the code complexity, and makes paFramesPerBufferUnspecified something achievable. It should also make non-interleaved buffers easier to handle.

tonetechnician commented:
Sorry, am resurrecting  this pull request from the dead. 

Has there been any more thought to merging this pull request? 

Seems like great progress was being made. Would love to have the callback method integrated as well! Also would be happy to do some tests and have a look over things

doesdev commented:
It would be an understatement to say I dropped the ball on this. Way back when the PR was opened it worked for me, should have been merged then. I'll need to run through things one more time to make sure that no changes within Node have caused this to not work. 

@tonetechnician If you wouldn't mind giving this branch a run through that would be helpful as well.

tonetechnician commented:
Cool! I'll test it and get back to you within the next couple days with a report. 

I think it would be worth migrating from NAN to NAPI for the ABI stability. Will see if I can try do something like that. Don't think it would be too difficult!

tonetechnician commented:
Cool. I actually managed to find some time quickly to check it out. 

This pull request is working on my end. I've tested the two added examples _pa_api_callback_sine.js_ and _pa_api_callback_wire.js_

My system:
__OS:__ Windows 10
__node__: v8.11.3
__npm__: v6.9.0


I do see within the build that some NAN stuff has been deprecated. Definitely think porting over to NAPI is a good option for a next step in this project. I may be able to handle this upon which I will make a pull request.

doesdev commented:
Unfortunately doesn't seem to be working on my Node version (11). I'll dig into why later (looks like maybe a removed interface at a glance).

tonetechnician commented:
Oh wow, that sounds pretty in depth. I need to get my NVM going and then can check it out as well. Is it to do with using NAN? I know on my node v8 I got a lot of build warnings saying NAN interfaces are deprecated. 

I really do want to port this over to NAPI, just need a bit of time - most likely in the next two weeks I'll be able to look into it. Just have some other deadlines coming up I need to get to first. I really want to use this in a project I'm working on, so I will definitely need to get to it sooner or later

doesdev commented:
Yeah, I'm sure it's a NAN issue. Definitely some TLC needed here. I never ramped up the project I intended this for so it kind of just went stale. I'd like to get the PA libs updated as well. I'm going to try and at least get what's in place working some time this week, and maybe start on the PA updates as well as merging this. That way if you are able to look into NAPI in a couple weeks you'll have a better base to work from.

tonetechnician commented:
All good man! Yeah, it would be amazing to get it updated to the latest PA lib! I guess the NAN stuff may take a bit more time. 

> That way if you are able to look into NAPI in a couple weeks you'll have a better base to work from.

Yes, would definitely be a better timeline to get the it working with the latest library and NAN, and then port. It shouldn't stop me from doing what I need to do though. I'll stay in touch here on the this pull request about when I start looking in to it

doesdev commented:
Okay, I've got the master branch reworked to accommodate API changes in v8 & Nan. That's in my `make-it-build` branch. All works on latest Nose version. I'm going to add some CI in to see how far back we can go on Node versions with the updates. Then I'll work on the callback PR.

tonetechnician commented:
You make quick sir! Nice one! I will check on my v8 node to check if it works. Will get my NVM up and running to speed up things up aswell.

Adding CI sounds great aswell, good call. 

I will be able to start looking into the NAPI port sometime in the next two weeks
